lesson_12/io/io_app/src/main/java/com/codedifferently/lesson12/factory/richhawkins/CsvDataLoader.java

```diff
@@ -0,0 +1,110 @@
+package com.codedifferently.lesson12.factory.richhawkins;
+
+import com.codedifferently.lesson12.factory.LibraryCsvDataLoader;
+import com.codedifferently.lesson12.models.CheckoutModel;
+import com.codedifferently.lesson12.models.LibraryDataModel;
+import com.codedifferently.lesson12.models.LibraryGuestModel;
+import com.codedifferently.lesson12.models.MediaItemModel;
+import com.opencsv.CSVReader;
+import com.opencsv.exceptions.CsvException;
+import java.io.FileReader;
+import java.io.IOException;
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import org.springframework.stereotype.Service;
+
+@Service
+public class CsvDataLoader implements LibraryCsvDataLoader {
+  @Override
+  public LibraryDataModel loadData() {
+    var model = new LibraryDataModel();
+    model.mediaItems =
+        readMediaItems(
+            "/workspaces/code-differently-24-q1/lesson_12/io/io_app/src/main/resources/csv/media_items.csv");
```

Comment:
You need to grab the files from an app relative path. Check out what I did in `LibraryJsonDataLoader` to solve this.

Reply:
Thanks, fixed this and a couple other things, all tests now pass!